crates/adapters/src/transport/durable/mod.rs

```diff
@@ -0,0 +1,192 @@
+///
+/// A "durable" data input transport divides input into steps numbered
+/// sequentially starting from zero.  Each step has durable content; that is, if
+/// it is read again, it will be the same as the first time.
+pub trait DurableInputEndpoint {
+    /// The range of steps that are available to be read.
+    fn steps(&self) -> AnyResult<Range<Step>>;
+
+    /// Opens the endpoint for reading steps and possibly adding them at the
+    /// end.  The reader passes data and status information to `consumer`.
+    fn open(&self, consumer: Box<dyn DurableInputConsumer>) -> Box<dyn DurableInputReader>;
+
+    /// Notifies the endpoint that steps less than `step` aren't needed
+    /// anymore.  It may optionally discard them.
+    fn expire(&self, _step: Step) {}
+}
+
+/// Reader and appender for [`DurableInputTransport`].
+///
+/// A `DurableInputReader` reads from a durable input transport step-by-step.
+/// It first reads the step specified in the first call to [`next`](Self::next),
+/// and advances as far as the highest step passed in any later call to
+/// [`next`](Self::next).  A reader can also cause new steps to be appended, by
+/// calling [`next`](Self::next) with a step number that does not yet exist.
+///
+/// Call [`DurableInputEndpoint::open`] to create a `DurableInputReader`.  The
+/// reader passes the data that it reads to the [consumer](DurableInputConsumer)
+/// passed to the `open` call.
+///
+/// The reader reads and appends asynchronously in separate threads.  Drop the
+/// reader to stop these threads and stop passing data to the consumer.
+///
+/// > 🚨️**Warning**: Only a single appender should be active at once for a given
+/// endpoint.
+pub trait DurableInputReader {
+    /// Requests that the endpoint reads steps up to and including `step`.  If
+    /// this is the first call for a given reader, then `step` will be the first
+    /// step read; otherwise, it requests continuing to read all of the steps up
+    /// to and including `step`.
+    ///
+    /// `step` has the following interpretation given that the input already
+    /// contains steps `first..last`:
+    ///
+    /// - If `step` is in `first..last`, then the reader will read up to and
+    /// including `step` and pass each step's data to the consumer.
+    ///
+    /// - If `step == last`, then the reader will wait for at least some data to
+    /// arrive, and then create `step` and pass it to the consumer.
+    ///
+    /// - If `step > last`, then the reader will create steps `last..step` with
+    /// any data that has already arrived (possibly none), and then wait for
+    /// some more data to arrive to populate `step`.  It will pass each of these
+    /// to the consumer in turn.
+    ///
+    /// - `step < first` will signal an error through [`InputConsumer::error`].
+    ///
+    /// Each call to `next` must specify a `step` at least as large as the
+    /// previous call.
+    fn next(&self, step: Step);
```

Comment:
ignore if this is stupid, but after reading the comment I wonder if it's better to supply a Range<u64> or similar here as an arg

Reply:
given that it's a reader and you're calling next() which is used by rust iterators, I'd have expected this to maybe return something, but it doesn't, maybe an example on how this is used would be good here